Police Tonfa - Sec Stun Baton alternative

Nakalon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This adds a new subtype of police baton, the Wooden Baton and places it up for purchase in the sectech vendor.

The idea was copying the deputy baton and change it so it feels better to play with and is more available.

The baton inherits most regular baton traits however the following is different:
- It displays attack animation on stamina attacks.
- It has less brute damage than the other batons (12). Having now the same brute (8) and throw damage (7) as the stun baton but, like the billy club, applies stam damage on harm as well (same as brute, 8)
- Deals more stamina damage (25) than the Deputy Baton (20) but considerably less than the Detective Baton (55)
- It has no cooldown as opposed to the Deputy Baton (10) and Detective Baton (20).
- Targeting different limbs has different outcomes:

Targeting the torso is the fastest way to stam crit.
Targeting the head will knock the target unconscious if his stamina damage is above 89.
Targeting the hands will disarm the person after 3 hits.
Targeting the legs makes the target trip.

A Huge **Huge** thanks to Haliris, without him losing like 3 hours explaining things to me this would have never happened.
Another Huge thanks to Absolucy who helped me with the cooldown and emoting.
Also a hearty thanks to Benbot, Tonty and Dejaku for giving me what I needed to figure out how to solve my problem!

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

It's meant to add an interesting alternative to the stun baton. Thus creating more RP opportunities for different types of characters (flavor).
It's meant to FEEL like combat without actually hurting the individual and meant to give a better conclusion to an arrest than a 2 tap stun baton if the suspect decides to fight.
It also would be difficult to type bait someone into stamcrit with this weapon.
By giving different limb targeting different outcomes more flavor is added to how the player decides to roleplay with the weapon.
Adds tactical depth as to which limb to target.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Concerns

Since I couldn't really force a custom emote I had to go through a different route of making a Knock-Out very clear. It may be a little too immersion-breaking right now and more on the funny side. I personally enjoy it but it is currently not at the level I wanted it to be.
I was unable to test both the Force Say and Knock-out at the same time for obvious reasons. It SHOULD work currently, but I have no proof other than the chatbox text disappearing as seen on the video.
📝Tweaks done, only thing left to do if possible is to make the emote show up in runechat how I originally intended, but this can be done at a later date.

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

🆕CURRENT STATE:
https://youtu.be/TroaLrOI5aU

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/2706f396-8b64-48c5-9091-bd00795c032f)

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/b085a134-bb59-4532-80a4-3740a20eed7a)

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/8eaa3e71-0ad9-4216-86d1-ca008b01c5c2)

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/603aaab2-023a-4593-bed4-18f67ec8cc6a)


</details>

## Changelog
:cl: PinkSuzuki feat Absolucy
Added: Police Tonfa
Added: Police Tonfa available at SecTech Vendor.
On this Weapon, limb attacks work as such on any intent but harm:
- Torso: 4 hits to Stam-Crit unarmored.
- Head: Knocks a target to Sleep if stamina damage is above a certain threshold, 5 hits to Sleep.
When Knocked-Out the target will be healed of all stamina damage.
Knock-Out has a duration of 8 seconds and a Cool-Down of 16 seconds (per target).
- Arms: Deals some stamina damage to the limb, but transfers the rest to the torso.
Disarm happens at 3 hits unarmored. 8 hits unarmored to Stam-Crit.
- Legs: Deals some stamina damage to the limb, but transfers the rest to the torso.
Attacking a Stam-Disabled leg trips the target. This ability has a 3-second cooldown (per target).
Tripping has a duration of 0.7 seconds.
Limb disabled at 2 hits unarmored. 7 hits unarmored to stam crit.
Harm Intent deals 8 brute damage and 8 stamina damage.

Code: Added a get_staminaloss proc for item/bodypart

/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Nakalon committed:
Farting

Adds farting to Bee Station

Nakalon committed:
Revert "Farting"

This reverts commit 68de4cea50a6c99c42c40710a226e61cdf8ab288.

Nakalon committed:
Adds Wooden Baton

Adds a wooden baton. Included in sectech vendor.

Nakalon committed:
Reduced Cooldown

Reduced cooldown to imitate regular combat.

tontyGH commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +326,16 @@
+	desc = "Favored by hot headed Security Officers who don't want to get in trouble with CentCom but still want to get that nostalgic feeling of beating some criminal scum upside the head with a chunk of wood."
+	force = 8
+	throwforce = 7
+	cooldown = 0
```
https://github.com/BeeStation/BeeStation-Hornet/assets/39193182/d4ab72e6-bfe1-4f55-88b6-80fb28b7ba7f

https://github.com/BeeStation/BeeStation-Hornet/assets/39193182/0b10573e-b0c3-41b1-8618-b58e5f107c2d

Nakalon replied:
This has been adressed in my last changes! Spent so long on this!

Nakalon committed:
Tripping & Disarm change

Changes how the weapon functions upon targeting different body parts.
Nerfs knockdown to the ground.

Nakalon committed:
Changes to Head targeting behaviour + Removing Stamheal on attack

Changes how the knock out blow to the head works.

Adds a sound, an emote and flavour text to it.
Also adds force say.

Reverts changes made to how new attacks would heal the targets stam damage.
Also adds a cooldown to a batons ability to sleep (20 seconds)

Added some lines to make code easier to read.

Nakalon committed:
Capitalization

Capitalizes the O in "Knock-Out!" to make it more like something you'd see in a fighting game.

Nakalon committed:
Balance Adjustements and New Proc

Adds some balance changes to ensure limb damage cannot be cheesed to reach stam crit faster.

Also adds a new get_staminaloss proc for limbs.

Nakalon committed:
Small Number Tweak

Missed some of the original values in recoding. They are now back.

Nakalon committed:
The Near-Prefection update.

Replaces the weapon cooldown by a per target.
Adds diminishing returns to random limb switching.
Cleans up the code some.
Adds the funny sound to tripping.
Adds proper logs and attack messages.
People on the ground can't be tripped anymore.
I think thats it.

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+			if((user.zone_selected == BODY_ZONE_HEAD))
+				target.apply_damage(12, STAMINA, BODY_ZONE_HEAD, def_check)
+
+				if(target.staminaloss > 89 && !target.has_status_effect(STATUS_EFFECT_SLEEPING) && (!sleep_cooldowns[target] || sleep_cooldowns[target] <= world.time))
```
In theory, since this only checks the cooldowns on this specific baton you can lock someone into sleep by hitting them while they're sleeping so that stam doesn't go under 89 and taking turns whacking the person in the head every 8 seconds with a buddy

Nakalon replied:
Incorrect! The cooldown applies PER TARGET, it is no longer tied to a weapon! Thust I lowered the cooldown form 20 to 16. (From per weapon to per target)

JixS4v replied:
It applies per target on a list on the weapon you doofus

JixS4v replied:
So it's still tied to the weapon

Nakalon replied:
Idk, blame Lucy, I'll talk to her

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+				else
+					target.apply_damage(4, STAMINA, BODY_ZONE_CHEST, def_check)
+
+				if(Ll.get_staminaloss() == 50 && CHECK_BITFIELD(target.mobility_flags, MOBILITY_STAND) && (!trip_cooldowns[target] || trip_cooldowns[target] <= world.time))
```
```suggestion
				if(Ll.get_staminaloss() == 50 && CHECK_BITFIELD(target.mobility_flags, MOBILITY_STAND) && (!trip_cooldowns[target] || COOLDOWN_FINISHED(src, trip_cooldowns[target]))
```
Better to use cooldown macros for readability.

Nakalon replied:
@Absolucy did this part of the code so she'd be better equip to answer this one!

Absolucy replied:
Honestly I'd never thought of using the macros on lists haha

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+					log_combat(user, target, "tripped", src)
+					target.visible_message(desc["visibletrip"], desc["localtrip"])
+					playsound(usr.loc, "sound/misc/slip.ogg", 30, 1)
+					trip_cooldowns[target] = world.time + 3 SECONDS
```
```suggestion
					COOLDOWN_START(src, trip_cooldowns[target], 30)
```

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+			if((user.zone_selected == BODY_ZONE_HEAD))
+				target.apply_damage(12, STAMINA, BODY_ZONE_HEAD, def_check)
+
+				if(target.staminaloss > 89 && !target.has_status_effect(STATUS_EFFECT_SLEEPING) && (!sleep_cooldowns[target] || sleep_cooldowns[target] <= world.time))
```
```suggestion
				if(target.staminaloss > 89 && !target.has_status_effect(STATUS_EFFECT_SLEEPING) && (!sleep_cooldowns[target] || COOLDOWN_FINISHED(src, sleep_cooldowns[target])))
```

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+				target.LAssailant = null
+			else
+				target.LAssailant = WEAKREF(user)
+			cooldown_check = world.time + cooldown
```
```suggestion
			COOLDOWN_START(src, cooldown_check, cooldown)
```

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+		if(!iscyborg(target))
+			return
+	else
+		if(cooldown_check <= world.time)
```
```suggestion
		if(COOLDOWN_FINISHED(src, cooldown_check))
```

itsmeow replied:
This was never implemented. It is the standard way to check a cooldown like that

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+						target.visible_message("<span class='emote'><b>[T]</b> [pick(list("falls unconscious.","falls limp like a bag of bricks.","falls to the ground, unresponsive.","lays down on the ground for a little nap.","got [T.p_their()] dome rung in."))]</span>")
+					else
+						target.visible_message("<span class='emote'><b>[T]</b> [pick(list("falls unconscious.","falls into a deep sleep.","was sent to dreamland.","closes [T.p_their()] and prepares for a little nap."))]</span>")
+					sleep_cooldowns[target] = world.time + 16 SECONDS
```
```suggestion
					COOLDOWN_START(src, sleep_cooldowns[target], 160)
```

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+				else
+					target.apply_damage(4, STAMINA, BODY_ZONE_CHEST, def_check)
+
+				if(Rl.get_staminaloss() == 50 && CHECK_BITFIELD(target.mobility_flags, MOBILITY_STAND) && (!trip_cooldowns[target] || trip_cooldowns[target] <= world.time))
```
```suggestion
				if(Rl.get_staminaloss() == 50 && CHECK_BITFIELD(target.mobility_flags, MOBILITY_STAND) && (!trip_cooldowns[target] || COOLDOWN_FINISHED(src, trip_cooldowns[target])))
```

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+					log_combat(user, target, "tripped", src)
+					playsound(usr.loc, "sound/misc/slip.ogg", 30, 1)
+					target.visible_message(desc["visibletrip"], desc["localtrip"])
+					trip_cooldowns[target] = world.time + 3 SECONDS
```
```suggestion
					COOLDOWN_START(src, trip_cooldowns[target], 30)
```

JixS4v commented on code/modules/surgery/bodyparts/bodyparts.dm:
What's the point of this proc, can't you just do bodypart.stamina_dam instead?

Nakalon replied:
I tried that and it didn't work! I also asked for a solution on #coding, and no one could give me another way to do this.

JixS4v commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,197 @@
+	if(iscyborg(target))
+		// We don't stun if we're on harm.
+		if (user.a_intent != INTENT_HARM)
+			if (affect_silicon)
```
Not sure if you copypasted this, but this affecting silicons is kind of insane since seccies already have a flash for this.

Nakalon replied:
Coppied this straight from the normal police baton, so whatever this does so does Deputy and Detective baton do!

Nakalon replied:
![image](https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/6e6c73d9-09ae-4b5d-a5cf-384ad021373d)
Also, this.

So no, it DOESN'T affect silicons.

JixS4v replied:
Might want to just take the code out, it's additional code for no reason

Nakalon replied:
It was there originally for some reason. I'm going to keep it for now until someone else comments over this, since you were wrong at first and I don't have the knowledge necessary to evaluate this well enough.

itsmeow replied:
If this baton is not intended to stun silicons, then you shouldn't include code to stun silicons, it's just extra code for no reason.

Nakalon committed:
New Naming and Sprite

A new name to the Baton, now its called OMNI Beater. Also has a new sprite and all that jazz

Nakalon committed:
Code Fix, Cooldown Fix, and Damage Buff

Fixes Cooldown - By Lucy
Tweaks the numbers a bit.

Nakalon committed:
oups, small number tweak

I made a mistake

JixS4v reviewed (approved):
Looks good, changes were adressed

Nakalon committed:
Name Change

Just a name change

itsmeow commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -326,6 +334,201 @@
+			return
+	else
+		if(COOLDOWN_FINISHED(src, cooldown_check))
+			if(ishuman(target))
```
this proc has issues with massive `if, else` chains.

I see you are using early returns, but maybe not enough or not in the right way.
Try to minimize indentation while maintaining the same logic.

If needed, you can also break parts into their own procs to reduce the indention level. You should not get over 2 indentation for the 'main code', otherwise the code is nigh unreadable.

Nakalon replied:
Yeah I'll probably need to redo all of this since I learned a lot since this, it was my first PR!

Nakalon replied:
I solved a bit i think

Nakalon replied:
I tried doing my absolute best, but I was unable to create procs for the limb damage.
I would need external help to do this as I am unable. So for now, this remains and if someone has something to say I'll have to request assistance

Nakalon committed:
Merge remote-tracking branch 'upstream/master' into PoliceBaton

Nakalon committed:
its good now

itsmeow commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -329,6 +337,185 @@
+			return
+		else
+			..()
+		return
```
This can be simplified.

```suggestion
		return ..()
```

itsmeow commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -329,6 +337,185 @@
+			..()
+		return
+	if(!isliving(target))
+		return
```
move this check before cyborg. cyborg is a type of living mob.

itsmeow commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -329,6 +337,185 @@
+		else
+			var/wait_desc = get_wait_description()
+			if (wait_desc)
+				to_chat(user, wait_desc)
```
Instead of having these else block at the end here, you can simplify and make the control flow more readable:

```
else if(cooldown_check > world.time)
    var/wait_desc = get_wait_description()
    if (wait_desc)
        to_chat(user, wait_desc)
    return
if(ishuman(target))
     //....
     if(H.check_shields(src, 0, "[user]'s [name]", MELEE_ATTACK))
          return
// check bodyzones ...
```

Nakalon replied:
I'll adress your concerns when im back to this project again!

Nakalon replied:
Adressed this!

Nakalon committed:
Kat simplification

Co-authored-by: itsmeow <itsmeow@itsmeow.dev>

Nakalon committed:
Merge remote-tracking branch 'upstream/master' into PoliceBaton

Nakalon committed:
Merge branch 'PoliceBaton' of https://github.com/Nakalon/BeeStation-Hornet into PoliceBaton

Nakalon committed:
Concerns adressed and buff

Nakalon committed:
added stamina damage on harm

Rukofamicom reviewed (approved):
Looks like all the requested changes requested were handled a while ago and I'd rather not lose this after all the work that went into it.
